src/lib/FlightTasks/tasks/Utility/StraightLine.cpp

```diff
@@ -0,0 +1,214 @@
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file StraightLine.cpp
+ */
+
+#include "StraightLine.hpp"
+#include <mathlib/mathlib.h>
+#include <float.h>
+
+#define VEL_ZERO_THRESHOLD 0.001f // Threshold to compare if the target velocity is zero
+#define DECELERATION_MAX 8.0f     // The vehicles maximum deceleration TODO check to create param
+
+using namespace matrix;
+
+StraightLine::StraightLine(ModuleParams *parent, const float &deltatime, const matrix::Vector3f &pos) :
+	ModuleParams(parent),
+	_deltatime(deltatime), _pos(pos)
+{
+
+}
+
+void StraightLine::generateSetpoints(matrix::Vector3f &position_setpoint, matrix::Vector3f &velocity_setpoint)
+{
+	// Check if target position has been reached
+	if (_is_target_reached || (_desired_speed_at_target < VEL_ZERO_THRESHOLD &&
+				   (_pos - _target).length() < NAV_ACC_RAD.get())) {
+		// Vehicle has reached target. Lock position
+		position_setpoint = _target;
+		velocity_setpoint = Vector3f(0.0f, 0.0f, 0.0f);
+		_is_target_reached = true;
+
+		return;
+	}
+
+	// unit vector in the direction of the straight line
+	Vector3f u_orig_to_target = (_target - _origin).unit_or_zero();
+	// vector from origin to current position
+	Vector3f orig_to_pos = _pos - _origin;
+	// current position projected perpendicularly onto desired line
+	Vector3f closest_pt_on_line = _origin + u_orig_to_target * (orig_to_pos * u_orig_to_target);
+	// previous velocity in the direction of the line
+	float speed_sp_prev = math::max(velocity_setpoint * u_orig_to_target, 0.0f);
+
+	// Calculate accelerating/decelerating distance depending on speed, speed at target and acceleration/deceleration
+	float acc_dec_distance = fabs(powf(_desired_speed, 2) - powf(_desired_speed_at_target, 2)) / 2.0f;
```

Comment:
I'm guessing `powf(x, 2)` takes noticable more cycles than x*x...
https://godbolt.org/g/wAqEwj
https://baptiste-wicht.com/posts/2017/09/cpp11-performance-tip-when-to-use-std-pow.html